src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -332,19 +315,35 @@ public SNIHandle CreateConnectionHandle(object callbackObject, string fullServer
                     break;
             }
 
+            if (isIntegratedSecurity)
+            {
+                string hostName = details.ServerName;
+                int connPort = details.Port;
+                string connInstanceName = details.InstanceName;
+
+                try
+                {
+                    spnBuffer = GetSqlServerSPN(hostName, (connPort >= 0 ? connPort.ToString() : connInstanceName));
+                }
+                catch (Exception e)
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.INVALID_PROV, SNICommon.ErrorSpnLookup, e);
+                }
+            }
+
             return sniHandle;
         }
 
-        private static byte[] GetSqlServerSPN(string hostNameOrAddress, string instanceName)
+        private static byte[] GetSqlServerSPN(string hostNameOrAddress, string portOrInstanceName)
         {
             Debug.Assert(!string.IsNullOrWhiteSpace(hostNameOrAddress));
             IPHostEntry hostEntry = Dns.GetHostEntry(hostNameOrAddress);
             string fullyQualifiedDomainName = hostEntry.HostName;
-            if (string.IsNullOrWhiteSpace(instanceName))
+            if (string.IsNullOrWhiteSpace(portOrInstanceName))
```

Comment:
If Named Pipes are used with default instance, the portOrInstanceName will be "" empty string. 
In that case you are trying to retrieve the SPN of 1433. 
What does Native SNI in this case? I think this case should return mssqlsvc/host.contoso.microsoft.com without any string appended. This will allow Kerberos auth to happen with the server SPN registered without the Port.